reconcile.go

```diff
@@ -234,54 +244,102 @@ func (a *Autopilot) pruneDeadServers() error {
 
 	state := a.GetState()
 
-	failed, voters, err := a.getFailedServers()
+	failed, vr, err := a.getFailedServers()
 	if err != nil || failed == nil {
 		return err
 	}
 
 	failed = a.promoter.FilterFailedServerRemovals(conf, state, failed)
 
-	// remove failed non voting servers
-	for _, srv := range failed.FailedNonVoters {
-		a.logger.Info("Attempting removal of failed server node", "id", srv.ID, "name", srv.Name, "address", srv.Address)
-		a.delegate.RemoveFailedServer(srv)
+	// Remove servers in order of increasing precedence
+	// Update the voter registry after each batch is processed
```

Comment:
Why are we changing the order here, e.g. before FailedNonVoters were considered first, now we're looking at them after the stale servers?

Reply:
I guess what I'm asking is: what's the basis for the "precedence" ordering you're using?

Reply:
I believe (and may well be totally wrong) it should be that stale servers are the lowest 'value', as to the application those servers don't exist (in its view). 

So first removing stale servers (non-voting and then voting) when possible, should be done before we attempt to remove failed servers (that the app does know about and could potentially become healthy again). 

A failed non-voter is still (most likely) a potential voter, a stale voter or non-voter won't be. 

Might be one which requires more thought/conversation :D

Reply:
I'm not sure that's true in Vault's case at least.  A stale server could simply be one whose echo message we haven't seen yet.  So it might be stale now, and in 10s it won't be.

Reply:
Vault would be the 'application' in this scenario, and Vault's change to say which servers should/shouldn't be part of the raft config is via `KnownServers` which it supplies. Naming both health and failed servers.

So if the server isn't in that collection of known servers, but *is* in the Raft configuration, then it's stale (from autopilot's perspective).. and needs to be removed. 

I may have been in this code for too long on my own, but that was how it seemed to me to work.

Reply:
So here in Vault's code a server is either healthy (alive) or failed (left) https://github.com/hashicorp/vault/blob/main/physical/raft/raft_autopilot.go#L428. If the server isn't in the map of servers, it's stale to Autopilot.

Reply:
I went back and forth on this for a while. There are tradeoffs here. If we remove stale first then we are moving things closer to the applications desired cluster config. However, this does involve messing with the raft config which is inherently a bit riskier than telling the application to remove a failed server. However, at least in Consul's usage, the request to remove the failed server will *almost always* result in the server being removed from the raft config very quickly after. 

For Consul's case the only time when one of those operations will not result in a raft config update is when the failed server isn't actually in the raft config yet. This seems like an extreme edge case and not something we need to explicitly account for.

So in the end I think we should follow two rules:

1. Deal with non-voters first as their removal shouldn't impact cluster stability.
2. Handle stale before failed so as to be moving things towards the applications desired server set.

The code as implemented does follow those rules. It would be great if you could add a comment about the ordering somewhere in here though (which I should have done when originally writing the library)

Reply:
I've added the rules you've mentioned in a comment in `pruneDeadServers` now, so hopefully the intent is documented to some extent.